Do not rethrow Content.Source exceptions

gregw opened the pull request:
Fix https://github.com/jetty/jetty.project/issues/13613. Due to a limitation in the implementation of try-with-resource, we should avoid rethrowing the same exception.  This PR makes a lazy copy of the exception in `Chunk.next(Chunk)`.

gregw committed:
Do not rethrow same exception

Fix #13613. Due to a limitation in the implementation of try-with-resource, we should avoid rethrowing the same exception.  This PR makes a copy of the exception.

gregw committed:
Do not rethrow same exception

Fix #13613. Due to a limitation in the implementation of try-with-resource, we should avoid rethrowing the same exception.  This PR makes a copy of the exception.

gregw committed:
Do not rethrow same exception

Fix #13613. Due to a limitation in the implementation of try-with-resource, we should avoid rethrowing the same exception.  This PR makes a copy of the exception.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -1106,7 +1107,32 @@ static Chunk next(Chunk chunk)
                 return null;
             if (Content.Chunk.isFailure(chunk))
-                return chunk.isLast() ? chunk : null;
+            {
```
I don't think this is the right place to handle this issue, I don't see anything wrong with `Content.Source.read()` always returning an error chunk that contains the exact same exception instance.

I think this should be handled at the place where the exception is thrown, like in `ContentSinkOutputStream`. Most (all?) places should already use `IO.rethrow()` so we can probably come up with some similar static helper, maybe like `Content.Chunk.throwIfFailure(Chunk)`?

We should also add some tests similar to `ContentSinkOutputStreamTest` for each codepath that throws a chunk's exception.

sbordet replied:
@lorban it is now clear that throwing the same instance is not well handled by the JDK and other libraries.

We cannot expect any user-code wrapper (such as `commons-fileupload`) to handle the case where the exception may be self suppressed or there is a loop.

I don't see any other choice than us throwing a different exception every time, or returning a new failure chunk every time.

I'd keep the code we added in #11759 and #13514, but I'm afraid we have to change the read side.

lorban replied:
> I don't see any other choice than us throwing a different exception every time, or returning a new failure chunk every time.

Yes, and my point is that we should throw a different exception every time, and NOT return a new failure chunk every time.

sbordet replied:
@lorban not sure what you mean here. The chunk is immutable, so we have to return a new chunk instance, just to change the exception instance inside.

User code (libraries and such) will just do `throw chunk.getFailure()` so we can't really ask them to wrap the failure into a new Throwable, can we?

joakime replied:
How about `throw chunk.getWrappedFailure(new IOException())` which just makes the existing `chunk.getFailure()` a cause of the one passed in?

lorban replied:
@sbordet I think we can ask users to not `throw chunk.getFailure()`, especially if we provide a helper.

If we don't and change the failure instance each time `getFailure()` is called, I expect difficulties tracing back a problem where an error chunk has been returned as the stack trace will be lost.

gregw replied:
But we have the Chunk.next() mechanism entirely to handle what the next chunk to be returned is.   If we read multiple times after a failure then we have to return different chunks if we want different exceptions.    So I think this is exactly the correct place to handle it.

sbordet replied:
I think we need to review `ContentSourceTransformer.read()`, which does not appear to use `next()` properly.

lorban replied:
@gregw if we do not lose the original cause (following the suggestion @sbordet made in `ExceptionUtil.copyOf()`) then I'm fine with putting the responsibility of making a copy in `Chunk.next()`.

We also have to revisit all our `Content.Source` implementations to make sure they properly call `Chunk.next()`.

gregw replied:
> @gregw if we do not lose the original cause (following the suggestion @sbordet made in `ExceptionUtil.copyOf()`) then I'm fine with putting the responsibility of making a copy in `Chunk.next()`.

I'm concerned that will still create problems.  But I'll try....
> 
> We also have to revisit all our `Content.Source` implementations to make sure they properly call `Chunk.next()`.

We should do that anyway!

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSinkOutputStream.java:
```diff
@@ -96,9 +97,13 @@ public void close(Callback callback) throws IOException
-        if (failed)
-            throw new IOException(failure.toString());
-        failed = true;
+        if (failed == null)
+            failed = failure;
+        else if (ExceptionUtil.areAssociated(failed, failure))
+            failure = IO.rethrow(ExceptionUtil.copyOf(failure));
+        else
+            failed.addSuppressed(failure);
+
```
I don't think we want to keep around a `Throwable` over a `boolean`, and I would not risk adding a large number of suppressed failures to a failure stored as a field.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -1106,7 +1107,32 @@ static Chunk next(Chunk chunk)
+                // handle persistent failure with lazy copy to avoid returning the same exception.
+                return new Empty()
+                {
+                    private transient Throwable failure;
```
Not `transient`.

I'd make it `final`, as the only thing to do with a failure chunk is to call `getFailure()`, so lazyness is not necessary.

I would just simplify this by calling `Chunk.from(copyOf(chunk.getFailure()))`.

gregw replied:
I think lazyness is required.    The normal case should be that we read the failure only once, so we do not want to create two exceptions in the normal case.  We should create the second exception only if there is a second read, and then create the third exception only if there is a third read etc.

sbordet replied:
I think we can do this with a boolean like we did in other places.
We don't need to return an `Empty` subclass the first time.
The subsequent times, we just call `Chunk.from(copyOf(chunk.getFailure())`.

gregw replied:
This is a static method, so we can't have a boolean here.   

Besides, the first chunk has already been returned and we are being asked for the next chunk here.  So the "boolean" is always `true`.  I.e. we have already returned the failure once and any use of the next chunk is the second usage (`true` already).

Any state we remember has to be in the Chunk that we return.   If we are passed in a non-transient failure Chunk, then we return a Chunk that IF AND ONLY IF it is used, will create another failure exception.

sbordet replied:
Ok re lazyness, but remove `transient` from the field modifiers.
`transient` is for Java serialization, and we don't care here.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ExceptionUtil.java:
```diff
@@ -76,6 +77,45 @@ public static <T extends Throwable> T as(Class<T> type, Throwable throwable) thr
+    {
+        if (failure == null)
+            return null;
+        for (Constructor<?> constructor : failure.getClass().getDeclaredConstructors())
+        {
+            Class<?>[] params = constructor.getParameterTypes();
+            if (params.length == 1 && params[0] == String.class)
+            {
+                try
+                {
+                    if (Modifier.isPublic(constructor.getModifiers()))
+                        return (Throwable)constructor.newInstance(failure.getMessage());
+                }
+                catch (Throwable ignored)
+                {
+                }
+                break;
+            }
+        }
```
I would rather try to find the `<ExceptionClass>(Throwable)` constructor first and failing that the parameterless constructor, and pass the original failure as cause or using `initCause()` on the new exception.
The `ISE` is thrown for suppression loops, not for cause loops.

Use `failure.getClass().getConstructor(Throwable.class)`, which returns only the `public` constructor with that signature.

gregw replied:
The problem with looking for a specific constructor with `getConstructor` is that it throws a `NoSuchMethod` exception if there is no such constructor.    Internally getConstructor(types...) loops through the constructors, so there is no efficiency difference.

We do not want to use pass the original failure as a cause, as the whole point is to avoid exception loops.  As the javadoc of this method says: without cause or suppressed exceptions.

sbordet replied:
> The problem with looking for a specific constructor with getConstructor is that it throws a NoSuchMethod exception if there is no such constructor. Internally getConstructor(types...) loops through the constructors, so there is no efficiency difference.

I don't see the problem?
Also, why copy/pasting JDK code, when the JDK does it for us in a 1 line?

> We do not want to use pass the original failure as a cause

But we don't want to lose the original cause, which has no problem being in a circular loop; only suppression does, and that's why we need to copy, but without losing the cause.

gregw replied:
The problem is that we do not want to create two exceptions and two stack traces for every Throwable that might have some special constructor.

gregw committed:
Updates from review

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/12.1.x/13613-do-not-rethrow

gregw committed:
updates from review

gregw committed:
updated tests to not expect same failure instance

gregw committed:
walk through supertypes

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ExceptionUtil.java:
```diff
@@ -76,6 +77,68 @@ public static <T extends Throwable> T as(Class<T> type, Throwable throwable) thr
+        Throwable cause = failure.getCause();;
+
+        Class<? extends Throwable> failureType = failure.getClass();
+        while (failureType != Exception.class && failureType != RuntimeException.class && failureType != Error.class && failureType != Throwable.class)
```
We can do these 2 pages of code with 2 lines of calling the JDK methods.
I don't see the reason to not call the JDK methods?
We should just call `failureType.getConstructor(String.class)` and then `initCause()`.
Am I missing something?

gregw committed:
Do not copy exception. Always throw a new IOException

sbordet committed:
Updates from reviews and fixed failing tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
EE9's AsyncContentProducer performs one more read in case of read failures.

Now that a new Chunk with a new failure is returned, retain the failure from
the second read only if the second failure's cause is different from the
failure of the first read.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/12.1.x/13613-do-not-rethrow

sbordet committed:
Updates from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
